modules/message-lane/src/lib.rs

```diff
@@ -90,19 +159,237 @@ decl_module! {
 		/// Deposit one of this module's events by using the default implementation.
 		fn deposit_event() = default;
 
+		/// Block initialization.
+		fn on_initialize(now: T::BlockNumber) -> Weight {
+			let mut dispatcher = T::QueuedMessageDispatch::with_any_weight();
+			let (queued_messages_dispatched, weight_spent_on_queued_dispatch) = dispatcher.dispatch();
+			if queued_messages_dispatched != 0 {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Spent {} weight on {} queued messages from on_initialize()",
+					weight_spent_on_queued_dispatch,
+					queued_messages_dispatched,
+				);
+			}
+
+			weight_spent_on_queued_dispatch
+		}
+
 		/// Send message over lane.
 		#[weight = 0] // TODO: update me (https://github.com/paritytech/parity-bridges-common/issues/78)
 		pub fn send_message(
 			origin,
 			lane_id: LaneId,
 			payload: T::Payload,
-		) {
-			let _ = ensure_signed(origin)?;
+			delivery_and_dispatch_fee: T::MessageFee,
+		) -> DispatchResult {
+			let submitter = ensure_signed(origin)?;
+
+			// let's first check if message can be delivered to target chain
+			T::TargetHeaderChain::verify_message(&payload).map_err(|err| {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Message to lane {:?} is rejected by target chain: {:?}",
+					lane_id,
+					err,
+				);
+
+				Error::<T, I>::MessageRejectedByChainVerifier
+			})?;
+
+			// now let's enforce any additional lane rules
+			T::LaneMessageVerifier::verify_message(
+				&submitter,
+				&delivery_and_dispatch_fee,
+				&lane_id,
+				&payload,
+			).map_err(|err| {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Message to lane {:?} is rejected by lane verifier: {:?}",
+					lane_id,
+					err,
+				);
+
+				Error::<T, I>::MessageRejectedByLaneVerifier
+			})?;
+
+			// let's withdraw delivery and dispatch fee from submitter
+			T::MessageDeliveryAndDispatchPayment::pay_delivery_and_dispatch_fee(
```

Comment:
What is supposed to happen with the fee here? I feel like this is too abstracted from the pallet. IMHO we should import `Currency` trait here and:
1. When someone is sending a message we should `lock` funds for `delivery_and_dispatch_fee`.
2. When the message is delivered (or processed - not sure), we should transfer the fee to the relayer.

Right now it's not clear to me which layer is supposed to be responsible for fees and how to implement this trait. The method name says `pay` so the straightforward way would be to do transfer right away (but to whom?), and I don't see any hooks for actually releasing the `fee`.

Reply:
> Right now it's not clear to me which layer is supposed to be responsible for fees and how to implement this trait. The method name says pay so the straightforward way would be to do transfer right away (but to whom?), and I don't see any hooks for actually releasing the fee.

I've left this decision for future PRs - please see section 2 of "description/Some ideas for future PRs".

Re excess abstraction - I don't think that we need to directly use `Currency` (or `LockableCurrency`) anywhere. Why do you think that lock+transfer is the only possible implementation? I see multiple uses for custom logic here. E.g. we may reserve some free slots for relayer transactions instead of paying him directly. Or it may be some increased possibility in next validators elections...Changing pallet API has larger (development) costs than changing implementations, so imo we should leave most generic callbacks where possible.

Reply:
I feel the fees problem (and also similar `delivery_and_dispatch_fee` discussion earlier) is a tradeoff between:
1. Extensibility of the pallets
2. Auditability / Ease of use.

Basically if we abstract more stuff out of the pallet, we allow more sophisticated schemes as you've mentioned, but it's becoming harder and harder to actually understand the whole thing.
I feel that since fees (and therefore) incentives are so complicated already I would prefer this to be an integral part of the delivery pallet - the pallet without proper incentive scheme is basically useless and it's might be even dangerous to include it in a production runtime.
Auditing the pallet seems also much easier than auditing the pallet + the runtime glue code.

The ease of use argument applies to both:
-  runtime integration (more custom hooks to implement, vs just plugging an existing `Currency` pallet, most likely just `pallet_balances`).
- usage of the external API - i.e. requirement of calling an extra runtime API before submitting a message (i.e. getting `delivery_and_dispatch_fee` is really much worse than simply saying how much you're willing to overpay to get your message faster (including a `tip`)) - this is obviously a bit more tricky, since probably selecting the right `tip` requires knowing the target chain dynamics, but still it seems a bit more ergonomic to me.